### PR TITLE
Fix isInLink tracking and improve recording of transitive deps file: in locks

### PR DESCRIPTION
**Pull request by iarna**
* [ ] Repro test from https://npm.community/t/npm-install-for-package-with-local-dependency-fails/754/10

Also fixes: https://npm.community/t/issue-npm-install-with-local-packages-and-symlinks-enoent/518

### CODE REVIEW

File: lib/shrinkwrap.js [hunk 1]
```diff
@@ -110,11 +110,13 @@ function shrinkwrapDeps (deps, top, tree, seen) {
     var childIsOnlyDev = isOnlyDev(child)
     var pkginfo = deps[moduleName(child)] = {}
     var requested = getRequested(child) || child.package._requested || {}
+    var linked = child.isLink || child.isInLInk
```
Comment (christoffer-dropbox) [inline lib/shrinkwrap.js]: `child.isInLInk` should probably be `child.isInLink `, right?

**Review by suttitum — APPROVED** [review]

File: lib/shrinkwrap.js (hunk 1 above)
Comment (christoffer-dropbox) [inline lib/shrinkwrap.js]: > `var linked = child.isLink || child.isInLInk`

`child.isInLink`?

Comment (iarna) [reply lib/shrinkwrap.js]: `child.isLink` = "This module is a symlink in your node_modules

`child.isInLink` = "Some parent module of this module is a symlink"

Comment (larsgw) [reply lib/shrinkwrap.js]: There seems to be a typo with a capital i in Link in isInLink.

Comment (iarna) [reply lib/shrinkwrap.js]: Ah, good catch, thank you!

Comment (larsgw) [reply lib/shrinkwrap.js]: I think it was @christoffer-dropbox who caught it, I just wanted to clarify what I thought he meant.

**Comment by ElvishJerricco** [comment]
Is this going to merged? This seems to be required for `file:` dependencies to have their own package-lock.json files.

**Comment by ghost** [comment]
With this fix, it is possible to install local packages with dependencies. However, a package-lock.json is not created and an existing one has no effect for those dependencies:
https://npm.community/t/npm-install-does-not-install-transitive-dependencies-of-local-dependency/2264/2?u=cbuchacher

**Comment by johnrjj** [comment]
How can I help get this PR merged ?

**Comment by PixleProgrambetaxl** [comment]
+1

**Comment by ElvishJerricco** [comment]
I would also like to help get this merged. Should I just look at why it's failing CI?

**Comment by jtwebman** [comment]
@zkat What do we need to do to get this PR in the next npm release. We are having to run npm 5.6.0 with node 10 till this is fixed as you can't use `file:../any folder` in projects.

**Comment by jtwebman** [comment]
FYI, I rebased this branch on latest and ran it locally it does fix the issue and I didn't have any other issues with it. Will run it on my local for awhile to see how well it does on my other projects as well. Not sure if that is what @zkat means by needs testing. I don't have access in Travis CI to re-run but @iarna there are a few issues with the tests. https://travis-ci.com/npm/cli/jobs/145118084 Maybe this breaks something else.

**Comment by jtwebman** [comment]
I also can run npm test after rebasing on latest and get no failures for this branch.

**Comment by jtwebman** [comment]
Adding some tests now for this issue. Maybe that is what they want. See PR https://github.com/npm/cli/pull/86

**Comment by jtwebman** [comment]
@iarna If you want to pull the test from my PR and merge or rebase this on latest I can close my PR. Not trying to take credit just trying to get this merged soon.

**Comment by ghost** [comment]
I think this fix is not correct. Please see https://github.com/npm/cli/pull/40#issuecomment-422748462 above:
https://npm.community/t/npm-install-does-not-install-transitive-dependencies-of-local-dependency/2264

**Comment by johnrjj** [comment]
@iarna -- Anything we can do to land this PR?

**Comment by jtwebman** [comment]
@cbuchacher why should a package-lock be created. The current project should be the only place a package-lock file is created. If you want a package-lock for the folder dependency then you need to run npm install in that folder. This should act just like if you pulled it from npm in my mind.

**Comment by ghost** [comment]
@jtwebman Well, it would be awesome if `npm install file:/local/path/to/mypackage` would act just like if we pulled the same package from npm. Unfortunately, it is not implemented like that. Instead, npm creates a symlink or directory junction (on Windows) at `node_modules/mypackage` pointing to `/local/path/to/mypackage`, and it will install dependencies of `mypackage` in `/local/path/to/mypackage/node_modules`. If you ask me, this strange implementation is the root cause for the numerous issues with local package dependencies.

**Comment by eliOcs** [comment]
@cbuchacher at the same time it is helpful that it is implemented with a symlink because there is no need to reinstall to get code updates.

**Comment by RidgeA** [comment]
@cbuchacher yes, it's helpful
But, at least in my case (monorepo with several microservices and several common libraries), it breaks build (`npm i`) and I have to remove `package-lock.json` to avoid this. 
Without lock-file installation lasts up to 10 minutes.

At least would be nice to have an option to make a normal installation instead of symlinking...

I hope this PR will land soon... :-)

**Comment by fennibay** [comment]
For non-local packages, I have to do an `npm i` to get code updates anyway, and conceptually it would be nicer to have a single behavior for local and non-local packages. So I think like @cbuchacher that symlink implementation should be replaced with a copy implementation.

**Comment by ghost** [comment]
A new _Workspaces_ concept is in the works which may address some of these issues:
https://npm.community/t/npm-needs-a-more-robust-story-for-local-multi-module-development/1059

**Comment by baptistemarchand** [comment]
Hello! Any chance of this being merged soon?

edit: if somebody finds it useful, I created a minimal example showing the problem of npm install with local dependencies: https://github.com/baptistemarchand/npm-local-path-bug-repro

**Comment by chrisyorkston** [comment]
Any idea when this will likely be merged?
I am having related issues with local dependencies.

**Comment by rifler** [comment]
hmm, I see code from this pr in `latest` branch

for example:
https://github.com/npm/cli/blob/latest/lib/shrinkwrap.js#L114

**Comment by larsgw** [comment]
Code from this PR was merged as part of another PR, https://github.com/npm/cli/pull/86.

**Comment by iarna** [comment]
As @larsgw points out, this has actually been merged, so I'm gonna close this now. =D

**Comment by aaronArinder** [comment]
Thanks for this PR, @iarna: super, super useful. :+1:

**Comment by lkaratun** [comment]
Still having this issue on `npm@6.9.1-next.0` :(

**Comment by spaceneenja** [comment]
It's 2020 and I'm down this rabbit hole on 6.13.4 and I think I'm going to pull the trigger and switch to Yarn.

**Pull request closed without merging**